battle-service/src/main/java/com/company/battle/controllers/GameController.java

```diff
@@ -1,43 +1,81 @@
+import com.company.common.models.SettingsEntity;
+import jakarta.validation.Valid;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.web.bind.annotation.*;
 
 import java.util.List;
 import java.util.UUID;
+import java.util.stream.Collectors;
 
 @RestController
 @RequestMapping("api/v1/game")
+@RequiredArgsConstructor
 public class GameController {
 
+    @Autowired
+    private final GameService gameService;
+
+    private final GameMapper gameMapper = GameMapper.INSTANCE;
+
+    private final SettingsMapper settingsMapper = SettingsMapper.INSTANCE;
+
     @GetMapping("/{id}")
     public GameDto getGameById(@PathVariable UUID id) {
-        return null;
+        GameEntity gameEntity = gameService.getById(id);
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @GetMapping("/{link}")
+    @GetMapping("by-link/{link}")
     public GameDto getGameByLink(@PathVariable String link) {
-        return null;
+        GameEntity gameEntity = gameService.getByLink(link);
+        return gameMapper.toDTO(gameEntity);
+    }
+
+    @GetMapping("/public-games")
+    public List<GameDto> getPublicGames() {
+        List<GameEntity> gameEntities = gameService.getPublicGames();
+        return gameEntities.stream()
+                .map(gameMapper::toDTO)
+                .collect(Collectors.toList());
     }
 
-    @GetMapping
-    public List<GameDto> getActiveGames() {
-        return null;
+    @GetMapping("/find-game")
+    public GameDto findGame(@Valid @RequestBody SettingsDto settingsDto) {
+        SettingsEntity settingsEntity = settingsMapper.toEntity(settingsDto);
+        GameEntity gameEntity = gameService.findGame(settingsEntity);
+        return gameMapper.toDTO(gameEntity);
     }
 
+    //ToDo: change randomUUID to userId from authorization
     @PostMapping("/create")
-    public GameDto createGame(@RequestBody GameDto gameDto) {
-        return null;
+    public GameDto createGame(@Valid @RequestBody GameDto gameDto) {
+        GameEntity gameEntity = gameMapper.toEntity(gameDto);
+        gameEntity = gameService.create(gameEntity, UUID.randomUUID());
+        return gameMapper.toDTO(gameEntity);
     }
 
+    //ToDo: finish this method
     @PutMapping("/join/{link}")
-    public GameDto joinGame(@PathVariable String link) {
-        return null;
+    public GameDto joinGame(@PathVariable String link, @PathVariable UUID userId) {
+        GameEntity gameEntity = gameService.getByLink(link);
+        gameEntity = gameService.joinGame(gameEntity, userId);
+        return gameMapper.toDTO(gameEntity);
     }
 
-    @PutMapping("/leave")
-    public GameDto leaveGame(@RequestBody GameDto gameDto) {
-        return null;
+    //ToDo: change userId as parameter to userId from authorization
+    @PutMapping("/leave/{userId}")
+    public GameDto leaveGame(@Valid @RequestBody GameDto gameDto, @PathVariable UUID userId) {
```

Comment:
you don't need to put game dto here.
you can use
/game/{gameId}/leave
and get user id from authorization context(you don't have authorization so far, but it's fine)
`
UserAuthentication authentication = (UserAuthentication) SecurityContextHolder.getContext().getAuthentication();`

Reply:
I changed gameDto with gameId. As for the userId, I will change it to the proposed option later, when the authorization  works. For this, I left a comment above the method. For now, I'm just mocking userId